PC/launcher2.c

```diff
@@ -0,0 +1,2165 @@
+ * runtime to be launched. It also enables auto-install of versions when they
+ * are requested but no installation can be found.
+ */
+
+#define __STDC_WANT_LIB_EXT1__ 1
+
+#include <windows.h>
+#include <pathcch.h>
+#include <shlobj.h>
+#include <stdio.h>
+#include <stdbool.h>
+#include <tchar.h>
+
+#define MS_WINDOWS
+#include "patchlevel.h"
+
+#define MAXLEN PATHCCH_MAX_CCH
+#define MSGSIZE 1024
+
+#define RC_NO_STD_HANDLES   100
+#define RC_CREATE_PROCESS   101
+#define RC_BAD_VIRTUAL_PATH 102
+#define RC_NO_PYTHON        103
+#define RC_NO_MEMORY        104
+#define RC_NO_SCRIPT        105
+#define RC_NO_VENV_CFG      106
+#define RC_BAD_VENV_CFG     107
+#define RC_NO_COMMANDLINE   108
+#define RC_INTERNAL_ERROR   109
+#define RC_DUPLICATE_ITEM   110
+#define RC_INSTALLING       111
+
+static FILE * log_fp = NULL;
+
+void
+debug(wchar_t * format, ...)
+{
+    va_list va;
+
+    if (log_fp != NULL) {
+        wchar_t buffer[MAXLEN];
+        int r = 0;
+        va_start(va, format);
+        r = vswprintf_s(buffer, MAXLEN, format, va);
+        va_end(va);
+
+        if (r <= 0) {
+            return;
+        }
+        fputws(buffer, log_fp);
+        while (r && isspace(buffer[r])) {
+            buffer[r--] = L'\0';
+        }
+        if (buffer[0]) {
+            OutputDebugStringW(buffer);
+        }
+    }
+}
+
+
+void
+formatWinerror(int rc, wchar_t * message, int size)
+{
+    FormatMessageW(
+        FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS,
+        NULL, rc, MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
+        message, size, NULL);
+}
+
+
+void
+winerror(int err, wchar_t * format, ... )
+{
+    va_list va;
+    wchar_t message[MSGSIZE];
+    wchar_t win_message[MSGSIZE];
+    int len;
+
+    if (err == 0) {
+        err = GetLastError();
+    }
+
+    va_start(va, format);
+    len = _vsnwprintf_s(message, MSGSIZE, _TRUNCATE, format, va);
+    va_end(va);
+
+    formatWinerror(err, win_message, MSGSIZE);
+    if (len >= 0) {
+        _snwprintf_s(&message[len], MSGSIZE - len, _TRUNCATE, L": %s",
+                     win_message);
+    }
+
+#if !defined(_WINDOWS)
+    fwprintf(stderr, L"%s\n", message);
+#else
+    MessageBoxW(NULL, message, L"Python Launcher is sorry to say ...",
+               MB_OK);
+#endif
+}
+
+
+void
+error(wchar_t * format, ... )
+{
+    va_list va;
+    wchar_t message[MSGSIZE];
+
+    va_start(va, format);
+    _vsnwprintf_s(message, MSGSIZE, _TRUNCATE, format, va);
+    va_end(va);
+
+#if !defined(_WINDOWS)
+    fwprintf(stderr, L"%s\n", message);
+#else
+    MessageBoxW(NULL, message, L"Python Launcher is sorry to say ...",
+               MB_OK);
+#endif
+}
+
+
+USHORT
+_getNativeMachine()
+{
+    static USHORT _nativeMachine = IMAGE_FILE_MACHINE_UNKNOWN;
+    if (_nativeMachine == IMAGE_FILE_MACHINE_UNKNOWN) {
+        USHORT processMachine;
+        if (!IsWow64Process2(NULL, &processMachine, &_nativeMachine)) {
```

Comment:
Will that load on Windows 8?

Reply:
No, it won't. I'll have to do it dynamically for Win 8.1 users (Win 8 itself is EOL).

Reply:
According to the documentation, though I haven't checked myself, it's also absent in Windows 10 version 1507 which is still under extended support in the LTSB branch.

In the fix for this I'd have change the `typedef` to something like `using PIsWow64Process2 = decltype(&IsWow64Process2);`
It's not critical and may be called a matter of style, but I see no benefit from providing the types yourself.